tests/integrations/opentelemetry/test_sampler.py

```diff
@@ -0,0 +1,351 @@
+
+import sentry_sdk
+
+
+tracer = trace.get_tracer(__name__)
+
+
+@pytest.fixture()
+def init_sentry_with_potel(sentry_init):
+    def wrapped_sentry_init(*args, **kwargs):
+        kwargs.update(
+            {
+                "_experiments": {"otel_powered_performance": True},
+            }
+        )
+        sentry_init(*args, **kwargs)
+
+    return wrapped_sentry_init
+
+
+@pytest.mark.parametrize(
+    "traces_sample_rate, expected_num_of_envelopes",
+    [
+        # special case for testing, do not pass any traces_sample_rate to init() (the default traces_sample_rate=None will be used)
+        (-1, 0),
+        # traces_sample_rate=None means do not create new traces, and also do not continue incoming traces. So, no envelopes at all.
+        (None, 0),
+        # traces_sample_rate=0 means do not create new traces (0% of the requests), but continue incoming traces. So envelopes will be created only if there is an incoming trace.
+        (0, 0),
+        # traces_sample_rate=1 means create new traces for 100% of requests (and also continue incoming traces, of course).
+        (1, 2),
+    ],
+)
+def test_sampling_traces_sample_rate_0_or_100(
+    init_sentry_with_potel,
+    capture_envelopes,
+    traces_sample_rate,
+    expected_num_of_envelopes,
+):
+    kwargs = {}
+    if traces_sample_rate != -1:
+        kwargs["traces_sample_rate"] = traces_sample_rate
+
+    init_sentry_with_potel(**kwargs)
+
+    envelopes = capture_envelopes()
+
+    with sentry_sdk.start_span(description="request a"):
+        with sentry_sdk.start_span(description="cache a"):
+            with sentry_sdk.start_span(description="db a"):
+                ...
+
+    with sentry_sdk.start_span(description="request b"):
+        with sentry_sdk.start_span(description="cache b"):
+            with sentry_sdk.start_span(description="db b"):
+                ...
+
+    assert len(envelopes) == expected_num_of_envelopes
+
+    if expected_num_of_envelopes == 2:
+        (transaction_a, transaction_b) = [
+            envelope.items[0].payload.json for envelope in envelopes
+        ]
+
+        assert transaction_a["transaction"] == "request a"
+        assert transaction_b["transaction"] == "request b"
+
+        spans_a = transaction_a["spans"]
+        assert len(spans_a) == 2
+        assert spans_a[0]["description"] == "cache a"
+        assert spans_a[1]["description"] == "db a"
+        spans_b = transaction_b["spans"]
+        assert len(spans_b) == 2
+        assert spans_b[0]["description"] == "cache b"
+        assert spans_b[1]["description"] == "db b"
+
+
+def test_sampling_traces_sample_rate_50(init_sentry_with_potel, capture_envelopes):
+    init_sentry_with_potel(traces_sample_rate=0.5)
+
+    envelopes = capture_envelopes()
+
+    # Make sure random() always returns the same values
+    with mock.patch(
+        "sentry_sdk.integrations.opentelemetry.sampler.random",
+        side_effect=[0.7, 0.7, 0.7, 0.2, 0.2, 0.2],
```

Comment:
this one is not a good test, because we want the transaction to be complete and this is bypassing the `parent_sampled` logic testing for the child spans.

I would wrap each transaction individually in a `mock.patch` that only stubs the first root span random call, the children should then derive from the parent decision.

Reply:
I changed them to make it easier readable. One Note: random() will be called on each span, because we retrieve the parent_sampled boolean and cast it to a float and use it as a sample rate. But you are right, the test looked clumsy :-)

Reply:
ok yea it's better now, I just want to avoid stubbing the wrong thing because that kind of flow will never happen in reality. If `parent_sampled` is set, it will always be 1 and never 0.7 or 0.2